core/src/main/java/io/grpc/MethodDescriptor.java

```diff
@@ -234,7 +237,29 @@ public boolean isIdempotent() {
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1775")
   public MethodDescriptor<ReqT, RespT> withIdempotent(boolean idempotent) {
     return new MethodDescriptor<ReqT, RespT>(type, fullMethodName, requestMarshaller,
-        responseMarshaller, idempotent);
+        responseMarshaller, idempotent, safe);
+  }
+
+  /**
+   * Returns whether this method is safe.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1775")
+  public boolean isSafe() {
+    return safe;
+  }
+
+  /**
+   * Setting safe on the method. Works only with unary rpcs.
+   * The method will try to fetch cached responses if it's both safe and idempotent. This is best
+   * effort and may have performance impact if the method is not desgined to be cacheable.
+   *
+   * @param safe whether the method is safe.
+   * @return a new copy of MethodDescriptor.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1775")
+  public MethodDescriptor<ReqT, RespT> withSafe(boolean safe) {
+    return new MethodDescriptor<ReqT, RespT>(type, fullMethodName, requestMarshaller,
```

Comment:
If this only works with unary, maybe have a checkArgument on the Type being unary?